dotnet/samples/KernelSyntaxExamples/Example58_FunctionCalling.cs

```diff
@@ -0,0 +1,104 @@
+using RepoUtils;
+
+/**
+ * This example shows how to use OpenAI's function calling capability via the chat completions interface.
+ * For more information, see https://platform.openai.com/docs/guides/gpt/function-calling.
+ */
+
+// ReSharper disable once InconsistentNaming
+public static class Example58_FunctionCalling
+{
+    public static async Task RunAsync()
+    {
+        IKernel kernel = await InitializeKernelAsync();
+        var chatCompletion = kernel.GetService<IChatCompletion>();
+        var chatHistory = chatCompletion.CreateNewChat();
+
+        await CompleteChatWithFunctionsAsync("What day is today?", chatHistory, chatCompletion, kernel);
+
+        await CompleteChatWithFunctionsAsync("What computer tablets are available for under $200?", chatHistory, chatCompletion, kernel);
+    }
+
+    private static async Task<IKernel> InitializeKernelAsync()
+    {
+        // Create kernel with chat completions service
+        IKernel kernel = new KernelBuilder()
+            .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+            .WithOpenAIChatCompletionService(TestConfiguration.OpenAI.ChatModelId, TestConfiguration.OpenAI.ApiKey, serviceId: "chat")
+            //.WithAzureChatCompletionService(TestConfiguration.AzureOpenAI.ChatDeploymentName, TestConfiguration.AzureOpenAI.Endpoint, TestConfiguration.AzureOpenAI.ApiKey, serviceId: "chat")
+            .Build();
+
+        // Load functions to kernel
+        string folder = RepoFiles.SampleSkillsPath();
+        kernel.ImportSemanticSkillFromDirectory(folder, "SummarizeSkill");
+        kernel.ImportSemanticSkillFromDirectory(folder, "WriterSkill");
+        kernel.ImportSemanticSkillFromDirectory(folder, "FunSkill");
+        kernel.ImportSkill(new TimeSkill(), "TimeSkill");
+
+        await kernel.ImportAIPluginAsync("KlarnaShoppingPlugin", new Uri("https://www.klarna.com/.well-known/ai-plugin.json"), new OpenApiSkillExecutionParameters());
+
+        return kernel;
+    }
+
+    private static async Task CompleteChatWithFunctionsAsync(string ask, ChatHistory chatHistory, IChatCompletion chatCompletion, IKernel kernel)
+    {
+        Console.WriteLine($"User message: {ask}");
+        chatHistory.AddUserMessage(ask);
+
+        // Retrieve available functions from the kernel and add to request settings
+        ChatRequestSettings requestSettings = new();
+        requestSettings.Functions = kernel.Skills.GetFunctionsView();
+
+        // Send request
+        var chatResult = (await chatCompletion.GetChatCompletionsAsync(chatHistory, requestSettings))[0];
+
+        // Check for message response
+        var chatMessage = await chatResult.GetChatMessageAsync();
+        if (chatMessage.Content.IsNotNullOrEmpty())
+        {
+            Console.WriteLine(chatMessage.Content);
+        }
+
+        // Check for function response
+        var functionCall = chatResult.ModelResult.GetResult<ChatModelResult>().Choice.Message.FunctionCall;
```

Comment:
I would propose to start adding additional behavior to our "data"/DTO/model classes to avoid the [train wrecks](https://wiki.c2.com/?TrainWreck) - `var functionCall = chatResult.ModelResult.GetResult<ChatModelResult>().Choice.Message.FunctionCall`

An alternative/example could be `var functionCall = chatResult.GetFunctionCall()`.

Reply:
Yeah, I can add it to ChatResult since that lives in the OpenAI connector. I'm hesitate to add it to IChatResult since that is in the Abstractions package -- but then can cast the result similar to GetResult<T>